http/src/main/scala/org/http4s/blaze/http/http2/server/AbstractBodyReader.scala

```diff
@@ -0,0 +1,90 @@
+package org.http4s.blaze.http.http2.server
+
+import java.nio.ByteBuffer
+
+import org.http4s.blaze.http.BodyReader
+import org.http4s.blaze.http.http2.{DataFrame, HeadersFrame, StreamMessage}
+import org.http4s.blaze.http.http2.Http2Exception._
+import org.http4s.blaze.util.{BufferTools, Execution}
+import org.log4s.getLogger
+
+import scala.concurrent.Future
+import scala.util.{Failure, Success}
+
+// TODO: this could probably be shared with the client
+/** Base implementation of the HTTP/2 [[BodyReader]]
+  *
+  * @param streamId the HTTP/2 stream id associated with this reader
+  * @param length length of data expected, or -1 if unknown.
+  */
+private abstract class AbstractBodyReader(streamId: Int, length: Long) extends BodyReader {
+  import AbstractBodyReader._
+
+  private[this] var bytesRead = 0L
+  private[this] var finished = false
+
+  private[this] val lock: Object = this
```

Comment:
How does lock.synchronized over the `this` reference work?

Reply:
It basically just makes everything synchronize on `this`. I don't know where I developed this pattern, but I typically like to have an explicit object that is being locked on, but perhaps this is just added noise. I'll just use the more idiomatic ` synchronized {`.